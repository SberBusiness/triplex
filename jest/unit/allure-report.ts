import {JasmineAllureReporter} from 'allure-jasmine';
import {Status, TestResult} from 'allure-js-commons';

const reporter = new JasmineAllureReporter({
    resultsDir: 'tests/out/unit/allure/allure-results',
    testMapper: (result: TestResult) => {
        if (result.status == Status.SKIPPED) result.fullName = `(WAS SKIPPED) ${result.fullName}`;
        return result;
    },
});

// @ts-ignore
jasmine.getEnv().addReporter(reporter);

export const allure = reporter.getInterface();
