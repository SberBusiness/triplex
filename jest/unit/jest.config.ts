import type {Config} from '@jest/types';

// Конфигурация Jest для запуска unit тесов.
const config: Config.InitialOptions = {
  preset: 'ts-jest',
  rootDir: '../../',
  roots: [
    '<rootDir>/src/',
    '<rootDir>/tools/',
  ],
  setupFilesAfterEnv: [
    '<rootDir>/jest/jest.setup.ts',
    '<rootDir>/jest/unit/allure-report.ts',
  ],
  moduleDirectories: [
    'node_modules',
    'src'
  ],
  moduleNameMapper: {
    "@jest/(.*)": "<rootDir>/jest/$1.ts",
    "^@sberbusiness/triplex/(.*)": "<rootDir>/src/$1"
  },
  verbose: true,
};

export default config;