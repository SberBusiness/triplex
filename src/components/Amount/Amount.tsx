import React from 'react';
import {formatAmount} from '@sberbusiness/triplex/utils/amountUtils';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';

/** Длина форматированной (с отступами и разделителем) строки amount, начиная с которой будет уменьшен шрифт. */
const adaptiveAmountLength = 14;

/** Свойства компонента Amount. */
interface IAmountProps extends React.HTMLAttributes<HTMLSpanElement> {
    /** Значение суммы. */
    value: string;
    /** Количество знаков после запятой. */
    fractionLength?: 0 | 1 | 2 | 3 | 4;
    /** Сокращённое обозначение валюты. */
    currency?: string;
    /** Сообщение подсказки названия валюты. */
    currencyTitle?: string;
    /** При большом количестве цифр уменьшает размер шрифта. */
    adaptive?: boolean;
    /** Тестовый атрибут. */
    dataTestId?: string;
}

/** Компонент отображения суммы. */
export const Amount: React.FC<IAmountProps> = ({
    className,
    value,
    fractionLength,
    currency,
    currencyTitle,
    adaptive,
    dataTestId,
    ...restProps
}) => {
    let formattedAmount = formatAmount(value, fractionLength);
    const classNames = classnames({'cssClass[amountAdaptive]': !!adaptive && formattedAmount.length >= adaptiveAmountLength}, className);

    if (formattedAmount[0] == '-') {
        // (Accessibility) Меняем дефис-минус на знак минуса для его озвучивания скрин-ридерами.
        formattedAmount = formattedAmount.replace('\u002D', '\u2212');
    }

    const renderCurrencyName = () => [
        '\u00A0',
        <span data-test-id={dataTestId && `${dataTestId}__currencyName`} title={currencyTitle} key="currencyName">
            {currency}
        </span>,
    ];

    return (
        <span className={classNames} {...restProps} data-tx={process.env.npm_package_version}>
            <span data-test-id={dataTestId && `${dataTestId}__amount`}>{formattedAmount}</span>
            {currency && renderCurrencyName()}
        </span>
    );
};

Amount.displayName = 'Amount';
