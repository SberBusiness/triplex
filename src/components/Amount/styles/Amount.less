@import 'src/styles/style';

.amountAdaptive {
    .font-size(@font-size-10);
}
