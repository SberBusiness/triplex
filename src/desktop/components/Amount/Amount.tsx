import {formatAmount} from '@sberbusiness/triplex/desktop/utils/amountUtils';
import {classnames} from '@sberbusiness/triplex/common/utils/classnames/classnames';
import * as React from 'react';

/** Длина форматированной (с отступами и разделителем) строки amount, начиная с которой будет уменьшен шрифт. */
const adaptiveAmountLength = 14;

/**
 * Свойства компонента Amount.
 *
 * @prop {string} value Значение суммы.
 * @param {number} [fractionLength] Количество знаков после запятой.
 * @prop {string} [currency] Сокращённое обозначение валюты.
 * @prop {string} [currencyTitle] Сообщение подсказки названия валюты.
 * @prop {boolean} [adaptive] При большом количестве цифр уменьшает размер шрифта.
 * @prop {string} [dataTestId] Тестовый атрибут.
 */
interface IAmountProps extends React.HTMLAttributes<HTMLSpanElement> {
    value: string;
    fractionLength?: 0 | 4;
    currency?: string;
    currencyTitle?: string;
    adaptive?: boolean;
    dataTestId?: string;
}

/** Компонент отображения суммы. */
export const Amount: React.FC<IAmountProps> = ({
    className,
    value,
    fractionLength,
    currency,
    currencyTitle,
    adaptive,
    dataTestId,
    ...restProps
}) => {
    const formattedAmount = formatAmount(value, fractionLength);
    let amountClassName: string | undefined;

    if (adaptive && formattedAmount.length >= adaptiveAmountLength) {
        amountClassName = 'cssClass[amountAdaptive]';
    }

    const renderCurrency = () => [
        '\u00A0',
        <span data-test-id={dataTestId && `${dataTestId}__currencyName`} title={currencyTitle} key="currencyName">
            {currency}
        </span>,
    ];

    return (
        <span {...restProps} className={classnames(className, amountClassName)}>
            <span data-test-id={dataTestId && `${dataTestId}__amount`}>{formattedAmount}</span>
            {currency && renderCurrency()}
        </span>
    );
};

Amount.displayName = 'Amount';
