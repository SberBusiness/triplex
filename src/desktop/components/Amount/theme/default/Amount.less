@import '../../../../common/theme/default';

.amountAdaptive {
    .font-size(@font-size-10);
}
