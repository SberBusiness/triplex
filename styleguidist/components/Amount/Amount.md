```jsx
import {Row} from '@sberbusiness/triplex/components/Row/Row';
import {Col} from '@sberbusiness/triplex/components/Col/Col';
import {Label} from '@sberbusiness/triplex/components/Label/Label';

const commonProps = {currency: 'RUB', currencyTitle: 'Российские рубли'};

const data = [
    {
        title: 'Обычный',
        element: <Amount value="8967452.3145" {...commonProps} />,
    },
    {
        title: 'Без копеек',
        element: <Amount value="8967452.31" fractionLength={0} {...commonProps} />,
    },
    {
        title: 'С 4 знаками после разделителя',
        element: <Amount key="2" value="8967452.31" fractionLength={4} {...commonProps} />,
    },
    {
        title: 'Положительный',
        element: <Amount value="+8967452.31" {...commonProps} />,
    },
    {
        title: 'Отрицательный',
        element: <Amount value="-8967452.31" {...commonProps} />,
    },
    {
        title: 'Без валюты',
        element: <Amount value="8967452.31" />,
    },
    {
        title: 'Большая сумма (50 миллиардов)',
        element: <Amount value="50000000000.31" {...commonProps} />,
    },
    {
        title: 'Адаптивная большая сумма (50 миллиардов)',
        element: <Amount value="50000000000.31" {...commonProps} adaptive />
    },
];

<>
    {data.map(({title, element}, index) => (
        <Row key={index}>
            <Col size={4}>
                <Label>
                    <Label.Text>{title}</Label.Text>
                </Label>
            </Col>
            <Col size={4}>{element}</Col>
        </Row>
    ))}
</>
```
