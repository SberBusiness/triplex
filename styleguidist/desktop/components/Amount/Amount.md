```jsx noeditor
import React from 'react';
import {AmountExample} from './AmountExample';
import ComponentPreview from '../../../common/components/ComponentPreview/ComponentPreview';

<>
    <ComponentPreview>
        <AmountExample />
    </ComponentPreview>
</>
```

### collapsed

```html { "file": "./AmountExample.tsx" }
```
