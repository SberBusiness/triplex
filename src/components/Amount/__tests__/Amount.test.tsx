import React from 'react';
import renderer from 'react-test-renderer';
import {Amount} from '../Amount';
import {allure} from '@jest/unit/allure-report';

describe('Amount', () => {
    beforeEach(() => {
        allure.feature('Amount');
    });

    it('renders correctly', () => {
        const component = <Amount value="-8964152,37" currency="RUB" currencyTitle="Российские рубли" dataTestId="amount" />;
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
