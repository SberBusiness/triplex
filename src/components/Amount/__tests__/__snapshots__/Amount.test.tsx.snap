// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amount renders correctly 1`] = `
<span
  className=""
>
  <span
    data-test-id="amount__amount"
  >
    -8 964 152,37
  </span>
   
  <span
    data-test-id="amount__currencyName"
    title="Российские рубли"
  >
    RUB
  </span>
</span>
`;
